protected/humhub/services/MigrationService.php

```diff
@@ -0,0 +1,258 @@
+    public const EVENT_BEFORE_MIGRATION = 'beforeMigration';
+
+    protected Module $module;
+    protected ?string $path;
+
+    /**
+     * @param Module $module
+     * @param array $config = ['path' => string|null]
+     */
+    public function __construct(Module $module, array $config = [])
+    {
+        $this->module = $module;
+
+        $path = $config['path'] ?? null;
+        unset($config['path']);
+
+        if ($path === null) {
+            $path = $module->basePath . '/migrations';
+
+            if (!is_dir(realpath($path))) {
+                Yii::debug("Module has no migrations directory.", $this->module->id);
+                $path = null;
+            }
+        } elseif (false === realpath($path)) {
+            $message = sprintf(
+                "Migration path '%s' for module '%s' is invalid!",
+                $path,
+                $this->module->getName()
+            );
+            Yii::error($message, $this->module->id);
+            throw new \RuntimeException($message);
+        }
+
+        $this->path = $path;
+
+        parent::__construct($config);
+    }
+
+    public function getModule(): Module
+    {
+        return $this->module;
+    }
+
+    public function getPath(): ?string
+    {
+        return $this->path;
+    }
+
+    public function migrateUp(string $action): ?bool
+    {
+        $result = new MigrationEvent([
+            'path' => $this->path,
+            'action' => $action,
+            'migration' => 'up',
+            'deactivateOnError' => true
+        ]);
+
+        $this->trigger(self::EVENT_BEFORE_MIGRATION, $result);
+
+        if ($result->path === null) {
+            return null;
+        }
+
+        Event::on(
+            MigrateController::class,
+            Controller::EVENT_AFTER_ACTION,
+            [
+                $this,
+                'onMigrationControllerAfterAction'
+            ],
+            $result
+        );
+
+        $result->data = MigrateController::webMigrateUp($result->path);
+
+        Event::off(
+            MigrateController::class,
+            Controller::EVENT_AFTER_ACTION,
+            [
+                $this,
+                'onMigrationControllerAfterAction'
+            ]
+        );
+
+        return $this->checkMigrationStatus($result);
+    }
+
+    /**
+     * Catches migration results.
+     */
+    public function onMigrationControllerAfterAction(ActionEvent $event)
+    {
+        if (!$event->sender instanceof MigrateController) {
+            return;
+        }
+
+        if (!$event->data instanceof MigrationEvent || $event->sender->migrationPath !== $event->data->path) {
+            return;
+        }
+
+        $event->data->result = $event->result ?? ExitCode::UNSPECIFIED_ERROR;
+    }
+
+    /**
+     * @param MigrationEvent $result
+     *
+     * @return bool
+     * @throws InvalidConfigException
+     * @throws Throwable
+     */
+    public function checkMigrationStatus(MigrationEvent $result): bool
```

Comment:
Doesn't need to be public. Was an automated code extraction in PhpStorm which defaults to `public`.

However, since the `MigrationService` could be extended from to incorporate additional functionality, `protected` might be an option? Or we make it `private` and if need arises, it can still be change to `protected`.

Reply:
I would say `private` when required make `protected`.